ast_lifter_410.ml

```diff
@@ -0,0 +1,1598 @@
+             ("Pexp_lazy", [this#lift_Parsetree_expression x0])
+       | Parsetree.Pexp_poly (x0, x1) ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_poly",
+               [this#lift_Parsetree_expression x0;
+               this#lift_option this#lift_Parsetree_core_type x1])
+       | Parsetree.Pexp_object x0 ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_object", [this#lift_Parsetree_class_structure x0])
+       | Parsetree.Pexp_newtype (x0, x1) ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_newtype",
+               [this#lift_loc this#string x0;
+               this#lift_Parsetree_expression x1])
+       | Parsetree.Pexp_pack x0 ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_pack", [this#lift_Parsetree_module_expr x0])
+       | Parsetree.Pexp_open (x0, x1) ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_open",
+               [this#lift_Parsetree_open_declaration x0;
+               this#lift_Parsetree_expression x1])
+       | Parsetree.Pexp_letop x0 ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_letop", [this#lift_Parsetree_letop x0])
+       | Parsetree.Pexp_extension x0 ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_extension", [this#lift_Parsetree_extension x0])
+       | Parsetree.Pexp_unreachable ->
+           this#constr "Ast_410.Parsetree.expression_desc"
+             ("Pexp_unreachable", []) : Parsetree.expression_desc ->
+                                          'res)
+    method lift_Parsetree_letop : Parsetree.letop -> 'res=
+      (fun
+         { Parsetree.let_ = let_; Parsetree.ands = ands;
+           Parsetree.body = body }
+         ->
+         this#record "Ast_410.Parsetree.letop"
+           [("let_", (this#lift_Parsetree_binding_op let_));
+           ("ands",
+             (this#list (List.map this#lift_Parsetree_binding_op ands)));
+           ("body", (this#lift_Parsetree_expression body))] : Parsetree.letop
+                                                                -> 'res)
+    method lift_Parsetree_binding_op : Parsetree.binding_op -> 'res=
+      (fun
+         { Parsetree.pbop_op = pbop_op;
+           Parsetree.pbop_pat = pbop_pat;
+           Parsetree.pbop_exp = pbop_exp;
+           Parsetree.pbop_loc = pbop_loc }
+         ->
+         this#record "Ast_410.Parsetree.binding_op"
+           [("pbop_op", (this#lift_loc this#string pbop_op));
+           ("pbop_pat", (this#lift_Parsetree_pattern pbop_pat));
+           ("pbop_exp", (this#lift_Parsetree_expression pbop_exp));
+           ("pbop_loc", (this#lift_Location_t pbop_loc))] : Parsetree.binding_op
+                                                              -> 'res)
+    method lift_direction_flag :
+      Migrate_parsetree__Ast_410.Asttypes.direction_flag -> 'res=
```

Comment:
`Migrate_parsetree__Ast_410` is already opened, it is not necessary to qualify the path here (this also applies to earlier versions).

More generally, the `Ast_4xx` string repeated so many times in the file is not necessary (we are dealing with a single version in each file, there can be no confusion).

Maybe the lifter methods should deal with prefixing the version number?? This can be dealt with later. In the meantime just removing the unnecessary qualifications should be enought.

Reply:
I just did a search-and-replace for this part of the diff. I'd prefer to treat further modifications separately from this PR, to stay focused here. Would you be willing to open an issue about this?

Reply:
Sure, let's mark this as resolved!

Reply:
The new things in `ast_lifter_410` are `method lift_Parsetree_functor_parameter`, and some places where I replaced `this#string` by `(this#lift_option this#string)`.

I realize all this excessive qualification probably makes it difficult to compare the `409` file to `410`, but one way around that is probably to simply do a search-and-replace in one before comparing with the other. This should be easier next time, after fixing the qualification, or ocaml-ppx/ppx will replace ppx_tools_versioned.

Reply:
Yes, I compared yours and ppx_tools implementation. It looks fine (a bit sad that such work had to be duplicated, but this is not a big overhead :P).

Reply:
Ha, I had assumed ppx_tools was dead since ppx_tools_versioned was created. Didn't know it's keeping up with the AST.